File: 961.n-repeated-element-in-size-2-n-array.0.kt
/*
 * @lc app=leetcode id=961 lang=kotlin
 *
 * [961] N-Repeated Element in Size 2N Array
 */
class Solution {
    fun repeatedNTimes(A: IntArray): Int =
        A.toList()
            .zip(A.toList().drop(1) + A.first())
            .dropWhile {(prev, curr) ->
                if (prev == curr) {
                    return prev
                }
                true
            }.firstOrNull().let { if (it == null) (if (A[0] == A[2]) A[0] else A[1]) else it.first }
}

File: 961.n-repeated-element-in-size-2-n-array.0.rs
/*
 * @lc app=leetcode id=961 lang=rust
 *
 * [961] N-Repeated Element in Size 2N Array
 */
impl Solution {
    pub fn repeated_n_times(a: Vec<i32>) -> i32 {
        let mut it = a.iter().cloned();
        let mut it2 = a.iter().cloned();
        it2.next();
        let ret = it.zip(it2).fold(-1, |ret, (p, c)| if (p == c) {p} else {ret});
        if (ret == -1) { if (a[0] == a[2] || a[0] == a[a.len() - 1]) {a[0]} else {a[1]} } else {ret}
    }
}
